pkg/runtime/wfengine/wfengine_test.go

```diff
@@ -368,22 +397,96 @@ func TestConcurrentActivityExecution(t *testing.T) {
+				if assert.NoError(t, err) {
+					assert.True(t, metadata.IsComplete())
+					assert.Equal(t, `["9","8","7","6","5","4","3","2","1","0"]`, metadata.SerializedOutput)
+
+					// Because all the activities run in parallel, they should complete very quickly
+					assert.Less(t, metadata.LastUpdatedAt.Sub(metadata.CreatedAt), 3*time.Second)
+				}
+			}
+		})
+	}
+}
+
+// TestContinueAsNewWorkflow verifies that a workflow can "continue-as-new" to restart itself with a new input.
+func TestContinueAsNewWorkflow(t *testing.T) {
+	r := task.NewTaskRegistry()
+	r.AddOrchestratorN("ContinueAsNewTest", func(ctx *task.OrchestrationContext) (any, error) {
+		var input int32
+		if err := ctx.GetInput(&input); err != nil {
+			return nil, err
+		}
 
-	id, err := client.ScheduleNewOrchestration(ctx, "ActivityFanOut")
-	if assert.NoError(t, err) {
-		metadata, err := client.WaitForOrchestrationCompletion(ctx, id)
-		if assert.NoError(t, err) {
-			assert.True(t, metadata.IsComplete())
-			assert.Equal(t, `["9","8","7","6","5","4","3","2","1","0"]`, metadata.SerializedOutput)
+		if input < 10 {
+			if err := ctx.CreateTimer(0).Await(nil); err != nil {
+				return nil, err
+			}
+			ctx.ContinueAsNew(input + 1)
+		}
+		return input, nil
+	})
+
+	ctx := context.Background()
+	client, engine := startEngine(ctx, r)
+	for _, opt := range GetTestOptions() {
+		t.Run(opt(engine), func(t *testing.T) {
+			id, err := client.ScheduleNewOrchestration(ctx, "ContinueAsNewTest", api.WithInput(0))
+			if assert.NoError(t, err) {
+				metadata, err := client.WaitForOrchestrationCompletion(ctx, id)
+				if assert.NoError(t, err) {
+					assert.True(t, metadata.IsComplete())
+					assert.Equal(t, `10`, metadata.SerializedOutput)
+				}
+			}
+		})
+	}
+}
 
-			// Because all the activities run in parallel, they should complete very quickly
-			assert.Less(t, metadata.LastUpdatedAt.Sub(metadata.CreatedAt), 3*time.Second)
+// TestRecreateCompletedWorkflow verifies that completed workflows can be restarted with new inputs externally.
+func TestRecreateCompletedWorkflow(t *testing.T) {
+	r := task.NewTaskRegistry()
+	r.AddOrchestratorN("EchoWorkflow", func(ctx *task.OrchestrationContext) (any, error) {
+		var input any
+		if err := ctx.GetInput(&input); err != nil {
+			return nil, err
 		}
+		return input, nil
+	})
+
+	ctx := context.Background()
+	client, engine := startEngine(ctx, r)
+	for _, opt := range GetTestOptions() {
+		t.Run(opt(engine), func(t *testing.T) {
+			// First workflow
+			var metadata *api.OrchestrationMetadata
+			id, err := client.ScheduleNewOrchestration(ctx, "EchoWorkflow", api.WithInput("echo!"))
+			if assert.NoError(t, err) {
+				if metadata, err = client.WaitForOrchestrationCompletion(ctx, id); assert.NoError(t, err) {
+					assert.True(t, metadata.IsComplete())
+					assert.Equal(t, `"echo!"`, metadata.SerializedOutput)
+				}
+			}
+
+			// Second workflow, using the same ID as the first but a different input
```

Comment:
I don't think we can test this here, but should we have tests (like in e2e scenarios) that check for behavior if the first instance is still running and you try to do this?

Reply:
I think we can test this here. I can create a workflow that sleeps forever, try to start it twice, and then verify that the second start fails.